request/do.go

```diff
@@ -52,21 +52,23 @@ func Do(requests int, quit <-chan struct{}, concurrency int, url string, timeout
 
 	var wg sync.WaitGroup
 
-	for i := 0; i < requests; i++ {
-		select {
-		case <-quit:
-			break
-		default:
+	go func() {
```

Comment:
goroutine is needed here, so we can in parallel collect records as they come in at line 72 (`collect(rec)`).

Reply:
`rec <- doRequest(url, timeout)` is already in a goroutine so it remains concurrent (otherwise it wouldn't have been working when the `for` wasn't wrapped in `func`), here this is just an anonymous func so we can `return`

Reply:
Calling `func() {}()` (not a goroutine, what you suggest) is the same as just inlining the body. So i'll compare with that.


`Do` :
```go
	rec := make(chan Record, requests) // buffered, has enough capacity for the entire loop

	for i := 0; i < requests; i++ {
		select {
		case <-quit:
			wg.Wait()
			close(rec)
			return collect(rec) // not called until quit is received. rec is not processed until then
		default:
		}
		acquire(sem, &wg)
		go func() {
			defer release(sem, &wg)
			rec <- doRequest(url, timeout)
		}()
	} // blocks here

	return collect(rec) // not called until the loop is exited. rec is not processed until then
```

`DoUntil`:
```go
	rec := make(chan Record) // non-buffered, will block until someone receives from it

	for {
		select {
		case <-quit:
			wg.Wait()
			close(rec)
			return collect(rec) // not called until quit is received. This will never happen because we'll block at doRequest() (unless quit is receive JUST after the first iteration but before the second).
		default:
		}
		acquire(sem, &wg)
		go func() {
			defer release(sem, &wg)
			rec <- doRequest(url, timeout) // blocks until someone received from rec (ie second iteration)
		}()
	}
```

Reply:
Ok, several things:

- I didn't notice you moved `wg.Wait()` and `close(rec)` inside `case <- quit`: this cannot work in `Do` as it'll never end until the global timeout is reached (we want to stop as soon as all requests are done). I think it should remain just before the final `return collect(rec)` and there should be only `return` in `case <- quit`

-  > Calling `func() {}()` (not a goroutine, what you suggest) is the same as just inlining the body
  Precisely, that is exactly what I meant: I expected the exact same behaviour as before (except the fixed `break` thing). That's why I suggested not to use a goroutine. Now try to use a goroutine after fixing the first point 🙃 

- I thought the `for` loop would be blocking only the time to throw the goroutines, not actually waiting for the request before going to the next iteration, this is indeed what we should aim for but I haven't figured why it isn't already the case 🤔

Reply:
Fixed in 20c36dab9a0511ba106bee2232871e791a85f772.
The outer goroutine is still needed because the for loop (or forever loop) waits for a free threads before launching an inner goroutine for the request itself.

It's not n threads and do the calls, but ♾️ threads with at most n concurrent and do the calls.

Reply:
Nothing's fixed 😬 It currently doesn't work.

Try running the following, the program never stops:
```sh
> go run cmd/runner/main.go -c 2 -r 10 http://localhost:9999\?delay\=500
```

Reply:
Works on my machine.

... just did not push